### PR TITLE
chore(explore): Update chart save to use API endpoints

### PR DESCRIPTION
**Pull request by codyml**
### SUMMARY

When a user saves a chart in Explore, the client sends one `POST /explore` request, then the server creates/updates the chart and optionally adds it to a new/existing dashboard as indicated.  This PR updates Explore's `SaveModal` to instead use v1 API endpoints to accomplish the same.  Here's the updated sequence of API calls:
- If user is creating a new dashboard: `POST /api/v1/dashboard/`
- If user is adding to a new dashboard: `PUT /api/v1/dashboard/{dashboard_id}`
- If user is saving as a new chart: `POST /api/v1/chart/`
  - If adding to new/existing dashboard: `{ dashboards: [..., dashboard_id] }`
- If user is overwriting an existing chart: `PUT /api/v1/chart/{chart_id}`
  - If adding to new/existing dashboard: `{ dashboards: [..., dashboardId] }`

In addition, this PR makes the Save and Save and Go to Dashboard buttons no longer trigger a page refresh.

### TESTING INSTRUCTIONS

Try performing the following tasks with and without proposed update and ensure that interface and user experience are identical (with the exception of there no longer being page refreshes):
- Create new chart from scratch, save as new chart, don't add to dashboard
- Create new chart from scratch, save as new chart, add to existing dashboard
- Create new chart from scratch, save as new chart, add to new dashboard
- Edit existing chart, overwrite existing chart, don't add to dashboard
- Edit existing chart, overwrite existing chart, add to existing dashboard
- Edit existing chart, overwrite existing chart, add to new dashboard
- Edit existing chart, save as new chart, don't add to dashboard
- Edit existing chart, save as new chart, add to existing dashboard
- Edit existing chart, save as new chart, add to new dashboard

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by michael-s-molina — COMMENTED** [review]
Thanks for the PR! I left some first-pass comments.

File: superset-frontend/src/explore/actions/saveModalActions.js
Comment (codyml) [reply superset-frontend/src/explore/actions/saveModalActions.js]: Changed to `parseInt` for clarity.

**Comment by kgabryje** [comment]
Can we get rid of page refresh when you overwrite the chart? I removed `window.location.assign(`/explore/?slice_id=${value.id}`);` from `SaveModal.tsx` and it seems to be working. The only problem is that "altered" pill doesn't disappear but I think that can be fixed easily by updating `explore.slice` (since slice holds the metadata of saved chart)

File: superset-frontend/src/explore/actions/saveModalActions.js
Comment (kgabryje) [inline superset-frontend/src/explore/actions/saveModalActions.js:92]:
```suggestion
      addSuccessToast(t('Chart %s has been saved', sliceName)),
```
That's the correct way of handling translations of strings with parameters. We can't use template literals in translations, because it would be impossible to use the params correctly if target language had different sentence structure than English

File: superset-frontend/src/explore/actions/saveModalActions.js
Comment (kgabryje) [inline superset-frontend/src/explore/actions/saveModalActions.js:108]: `t('Dashboard %s just got created and chart %s was added to it', addedToDashboard.title, sliceName)`

File: superset-frontend/src/explore/actions/saveModalActions.js
Comment (codyml) [reply superset-frontend/src/explore/actions/saveModalActions.js]: Oops makes sense, thanks!

**Comment by codyml** [comment]
> Can we get rid of page refresh when you overwrite the chart? I removed `window.location.assign(`/explore/?slice_id=${value.id}`);` from `SaveModal.tsx` and it seems to be working. The only problem is that "altered" pill doesn't disappear but I think that can be fixed easily by updating `explore.slice` and/or `charts[chart_id].sliceFormData` (since slice holds the metadata of saved chart)
> 
> Ideally we could also prevent redirecting when saving as a new chart, but that might be trickier...

Working on this as a separate fix outside of this PR.

**Review by kgabryje — APPROVED** [review]
Tested again, looks good to me!
Let's wait for more approvals before merging

**Comment by EugeneTorap** [comment]
Hi @codyml @kgabryje @michael-s-molina 
Can we do `Transition to SQL Lab with React Router` next PR?

https://user-images.githubusercontent.com/29536522/177793452-38e9e271-1f69-4924-9f8f-5359efd03cfb.mov

**Review by geido — APPROVED** [review]
Could not find regressions. LGTM!

**Comment by codyml** [comment]
Holding for QA review

**Review by zhaoyongjie — COMMENTED** [review]
leave a couple of suggestions. lgtm

**Comment by codyml** [comment]
> Hi @codyml. I can only repro 1 issue that found found earlier (When overwriting a chart and selecting a different dashboard than the one it's currently on in the modal, the chart is removed from the dashboard it's currently on.).
> Have you fixed it or not and are you planning to fix it?

@andrey-zayats I thought I had fixed it but I guess not :( will check it out and let you know when it's resolved, thanks!

File: superset-frontend/src/explore/actions/saveModalActions.js
Comment (codyml) [reply superset-frontend/src/explore/actions/saveModalActions.js:163]: Oh thanks, had no idea!

File: superset-frontend/src/explore/actions/saveModalActions.js
Comment (codyml) [reply superset-frontend/src/explore/actions/saveModalActions.js:94]: I'm not sure why non-primitives are serialized, I was mostly just keeping what was there before and I confess I don't really understand what's happening with all of these function calls that generate the payload.  Are you sure it's safe to leave out all of those params?

File: superset-frontend/src/explore/ExplorePage.tsx
Comment (codyml) [reply superset-frontend/src/explore/ExplorePage.tsx:25]: @kgabryje I think this was your code.

File: superset-frontend/src/explore/ExplorePage.tsx
Comment (codyml) [reply superset-frontend/src/explore/ExplorePage.tsx:60]: @kgabryje

File: superset-frontend/src/explore/ExplorePage.tsx
Comment (kgabryje) [reply superset-frontend/src/explore/ExplorePage.tsx:25]: Yeah I wasn't aware of `isDefined`. Feel free to refactor 🙂

File: superset-frontend/src/explore/ExplorePage.tsx
Comment (codyml) [reply superset-frontend/src/explore/ExplorePage.tsx:60]: @zhaoyongjie @kgabryje Not sure I totally understand this suggestion – is there an advantage to using `isDefined` over `isNullish`?  Also, isn't the suggested code not equivalent to the original code because the original code throws if both `id` and `uid` are blank, while the suggested code throws if either are blank?

Comment (zhaoyongjie) [reply superset-frontend/src/explore/ExplorePage.tsx:60]: The same logic, only optimized for readability.

**Comment by jinghua-qa** [comment]
I have found an issue that when add a new chart from explore to the dashboard and apply native filter to the dashboard, the filter indicator is not appea

https://user-images.githubusercontent.com/81597121/178197199-a1691137-5c36-4baa-9fb5-724cbf22598f.mov

r

**Comment by jinghua-qa** [comment]
> I have found an issue that when add a new chart from explore to the dashboard and apply native filter to the dashboard, the filter indicator is not appear

Verified this issue is also exist in master, not related to this PR. So LGTM!

**Comment by michael-s-molina** [comment]
Hi @codyml. [This](https://github.com/apache/superset/pull/20676) fixes codecov CI error. You need to rebase your PR.

**Pull request merged**
